File: api/core/workflow/entities/nested_variable.py
"""
Nested Variable entities for workflow nodes.

This module provides data structures for defining nested/hierarchical variables
that can contain child variables, enabling complex object and array structures
in workflow node inputs and outputs.
"""

import re
from collections.abc import Sequence
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field, field_validator, model_validator

MAX_NESTING_DEPTH = 5


class NestedVariableType(StrEnum):
    """Supported types for nested variables."""

    STRING = "string"
    INTEGER = "integer"
    NUMBER = "number"
    BOOLEAN = "boolean"
    OBJECT = "object"
    FILE = "file"
    ARRAY_STRING = "array[string]"
    ARRAY_INTEGER = "array[integer]"
    ARRAY_NUMBER = "array[number]"
    ARRAY_OBJECT = "array[object]"
    ARRAY_BOOLEAN = "array[boolean]"
    ARRAY_FILE = "array[file]"

    def is_nestable(self) -> bool:
        """Check if this type supports child variables."""
        return self in (NestedVariableType.OBJECT, NestedVariableType.ARRAY_OBJECT)

    def is_array(self) -> bool:
        """Check if this is an array type."""
        return self.value.startswith("array[")


# Regex pattern for valid variable names: starts with letter, contains only alphanumeric and underscores
VARIABLE_NAME_PATTERN = re.compile(r"^[a-zA-Z][a-zA-Z0-9_]*$")


class NestedVariableDefinition(BaseModel):
    """
    Definition of a nested variable.

    Supports hierarchical structures where Object and Array<Object> types
    can contain child variable definitions.
    """

    name: str = Field(description="Variable name")
    type: NestedVariableType = Field(description="Variable type")
    required: bool = Field(default=False, description="Whether the variable is required")
    description: str = Field(default="", description="Variable description")
    default_value: Any = Field(default=None, description="Default value")
    children: Sequence["NestedVariableDefinition"] | None = Field(
        default=None,
        description="Child variable definitions, only for object and array[object] types",
    )

    @field_validator("name")
    @classmethod
    def validate_name(cls, v: str) -> str:
        """Validate that variable name follows the required pattern."""
        if not VARIABLE_NAME_PATTERN.match(v):
            raise ValueError(
                f"Variable name '{v}' must start with a letter and contain only alphanumeric characters and underscores"
            )
        return v

    @model_validator(mode="after")
    def validate_children_for_type(self) -> "NestedVariableDefinition":
        """Validate that children are only allowed for nestable types."""
        if self.children is not None and len(self.children) > 0:
            if not self.type.is_nestable():
                raise ValueError(f"Children are only allowed for 'object' and 'array[object]' types, got '{self.type}'")
            # Validate child variable names are unique within the same parent
            names = [child.name for child in self.children]
            if len(names) != len(set(names)):
                duplicates = [name for name in names if names.count(name) > 1]
                raise ValueError(
                    f"Child variable names must be unique within the same parent. Duplicates: {set(duplicates)}"
                )
        return self

    def get_max_depth(self) -> int:
        """Calculate the maximum nesting depth of this variable definition."""
        if not self.children:
            return 1
        return 1 + max(child.get_max_depth() for child in self.children)

    def validate_depth(self, max_depth: int = MAX_NESTING_DEPTH) -> list[str]:
        """
        Validate that nesting depth does not exceed the maximum.

        Returns:
            List of error messages, empty if validation passes.
        """
        errors: list[str] = []
        current_depth = self.get_max_depth()
        if current_depth > max_depth:
            errors.append(f"Maximum nesting depth of {max_depth} exceeded at '{self.name}' (depth: {current_depth})")
        return errors

    def __repr__(self) -> str:
        children_count = len(self.children) if self.children else 0
        return f"NestedVariableDefinition(name={self.name!r}, type={self.type}, children={children_count})"


# Rebuild model to allow recursive type references
NestedVariableDefinition.model_rebuild()

File: web/types/workflow/nested-variable.ts
/**
 * Nested Variable Type Definitions
 *
 * This module defines TypeScript types for nested variable support in Dify workflows.
 * Nested variables allow users to define complex data structures with hierarchical
 * relationships, supporting Object and Array<Object> types with child variables.
 *
 * @see Requirements 1.3, 1.4 - Type support for child variables
 */

/**
 * Supported types for nested variables.
 * Includes primitive types, array types, and nestable types (object, array[object]).
 */
export enum NestedVariableType {
  // Primitive types
  STRING = 'string',
  INTEGER = 'integer',
  NUMBER = 'number',
  BOOLEAN = 'boolean',
  FILE = 'file',

  // Nestable types - can have children
  OBJECT = 'object',

  // Array types
  ARRAY_STRING = 'array[string]',
  ARRAY_INTEGER = 'array[integer]',
  ARRAY_NUMBER = 'array[number]',
  ARRAY_BOOLEAN = 'array[boolean]',
  ARRAY_FILE = 'array[file]',
  ARRAY_OBJECT = 'array[object]',
}

/**
 * Maximum allowed nesting depth for nested variables.
 * Prevents overly complex structures that could impact performance.
 */
export const MAX_NESTING_DEPTH = 5

/**
 * Regular expression pattern for validating variable names.
 * Names must start with a letter and contain only alphanumeric characters and underscores.
 */
export const VARIABLE_NAME_PATTERN = /^[a-zA-Z]\w*$/

/**
 * Checks if a variable type supports nested children.
 * Only 'object' and 'array[object]' types can have child variables.
 */
export function isNestableType(type: NestedVariableType): boolean {
  return type === NestedVariableType.OBJECT || type === NestedVariableType.ARRAY_OBJECT
}

/**
 * Checks if a variable type is an array type.
 */
export function isArrayType(type: NestedVariableType): boolean {
  return type.startsWith('array[')
}

/**
 * Validates a variable name against the naming pattern.
 * @param name - The variable name to validate
 * @returns true if the name is valid, false otherwise
 */
export function isValidVariableName(name: string): boolean {
  return VARIABLE_NAME_PATTERN.test(name)
}

/**
 * Definition of a nested variable.
 * Supports recursive structure for Object and Array<Object> types.
 */
export type NestedVariableDefinition = {
  /** Variable name - must match VARIABLE_NAME_PATTERN */
  name: string
  /** Variable type */
  type: NestedVariableType
  /** Whether the variable is required */
  required: boolean
  /** Optional description of the variable */
  description?: string
  /** Optional default value */
  defaultValue?: unknown
  /**
   * Child variable definitions.
   * Only valid for 'object' and 'array[object]' types.
   */
  children?: NestedVariableDefinition[]
}

/**
 * Enhanced variable selector that supports nested paths.
 * Used to reference variables from upstream nodes with nested access.
 */
export type EnhancedVariableSelector = {
  /** Variable reference string */
  variable: string
  /** Variable selector path - array of path segments */
  valueSelector: string[]
}

/**
 * Node input definition with nested variable support.
 */
export type NodeInputDefinition = {
  /** Input name */
  name: string
  /** Input type */
  type: NestedVariableType
  /** Whether the input is required */
  required: boolean
  /** Optional description */
  description?: string
  /** Variable selector for referencing upstream variables */
  variableSelector?: EnhancedVariableSelector
  /** Child variable definitions for nested types */
  children?: NestedVariableDefinition[]
  /** Default value */
  defaultValue?: unknown
}

/**
 * Node output definition with nested variable support.
 */
export type NodeOutputDefinition = {
  /** Output name */
  name: string
  /** Output type */
  type: NestedVariableType
  /** Optional description */
  description?: string
  /** Child variable definitions for nested types */
  children?: NestedVariableDefinition[]
}

/**
 * Props for the NestedVariableEditor component.
 */
export type NestedVariableEditorProps = {
  /** Current variable definitions */
  value: NestedVariableDefinition[]
  /** Callback when definitions change */
  onChange: (value: NestedVariableDefinition[]) => void
  /** Maximum nesting depth allowed (default: MAX_NESTING_DEPTH) */
  maxDepth?: number
  /** Whether the editor is disabled */
  disabled?: boolean
  /** Current nesting depth (used internally for recursion) */
  currentDepth?: number
  /** Parent path for generating unique keys (used internally) */
  parentPath?: string
}

/**
 * Validation error for nested variables.
 */
export type NestedVariableValidationError = {
  /** Error code */
  code: NestedVariableErrorCode
  /** Human-readable error message */
  message: string
  /** Path to the problematic variable (e.g., 'user.profile.age') */
  path?: string
  /** Expected type (for type mismatch errors) */
  expected?: string
  /** Actual type (for type mismatch errors) */
  actual?: string
}

/**
 * Error codes for nested variable validation.
 */
export enum NestedVariableErrorCode {
  INVALID_VARIABLE_NAME = 'INVALID_VARIABLE_NAME',
  DUPLICATE_CHILD_NAME = 'DUPLICATE_CHILD_NAME',
  MAX_DEPTH_EXCEEDED = 'MAX_DEPTH_EXCEEDED',
  INVALID_CHILDREN_TYPE = 'INVALID_CHILDREN_TYPE',
  REQUIRED_FIELD_MISSING = 'REQUIRED_FIELD_MISSING',
  TYPE_MISMATCH = 'TYPE_MISMATCH',
}

/**
 * Type guard to check if a value is a NestedVariableDefinition.
 */
export function isNestedVariableDefinition(value: unknown): value is NestedVariableDefinition {
  if (typeof value !== 'object' || value === null)
    return false

  const obj = value as Record<string, unknown>
  return (
    typeof obj.name === 'string'
    && typeof obj.type === 'string'
    && Object.values(NestedVariableType).includes(obj.type as NestedVariableType)
    && typeof obj.required === 'boolean'
  )
}

/**
 * Calculates the maximum nesting depth of a variable definition.
 * @param definition - The variable definition to analyze
 * @returns The maximum depth (1 for leaf nodes)
 */
export function getMaxDepth(definition: NestedVariableDefinition): number {
  if (!definition.children || definition.children.length === 0)
    return 1

  return 1 + Math.max(...definition.children.map(getMaxDepth))
}

/**
 * Gets the full path string from a variable selector.
 * @param selector - The enhanced variable selector
 * @returns Dot-notation path string
 */
export function getFullPath(selector: EnhancedVariableSelector): string {
  return selector.valueSelector.join('.')
}
